bindings/c/src/error.rs

```diff
@@ -81,3 +71,61 @@ impl opendal_error {
         }
     }
 }
+
+/// \brief The opendal error type for C binding, containing an error code and corresponding error
+/// message.
+///
+/// The normal operations returns a pointer to the opendal_error, and the **nullptr normally
+/// represents no error has taken placed**. If any error has taken place, the caller should check
+/// the error code and print the error message.
+///
+/// The error code is represented in opendal_code, which is a enum on different type of errors.
+/// The error messages is represented in opendal_bytes, which is a non-null terminated byte array.
+///
+/// \note 1. The error message is on heap, so the error needs to be freed by the caller, by calling
+///       opendal_error_free. 2. The error message is not null terminated, so the caller should
+///       never use "%s" to print the error message.
+///
+/// @see opendal_code
+/// @see opendal_bytes
+/// @see opendal_error_free
+#[repr(C)]
+pub struct opendal_error {
+    code: opendal_code,
+    message: opendal_bytes,
+}
+
+impl opendal_error {
+    // The caller should sink the error to heap memory and return the pointer
+    // that will not be freed by rustc
+    pub(crate) fn from_opendal_error(error: od::Error) -> Self {
```

Comment:
opendal_error and opendal_internal_error is a bit confusing. How about use:

- `opendal_error`: the error we exported to C
- `raw_error`: the error wrapper for rust (don't need to have opendal prefix).